### PR TITLE
feat(lib): Add DynamicListTerraformIterator with support for lists or sets with objects that have properties that are only known at apply time

### LINKED ISSUE #2178: how to implement AcmCertificateValidation in typescript?
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I don't believe it is currently possible (CDKTF 0.13.0) in Typescript to generate a ACM Certificate because of the complex inputs needed in an [acm_certificate_validation](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate_validation#dns-validation-with-route-53) operation. Setting up a certificate is a fundamental piece before creating more infrastructure.

Please let me know if I have somehow missed an example that shows it is possible to validate a certificate with DNS validation.

There is some discussion around [iterators](https://www.terraform.io/cdktf/concepts/iterators). It seems to me the biggest problem is trying to re-use `for_each` from HCL instead of providing a [proper iterator](https://github.com/hashicorp/terraform-cdk/issues/2001#issuecomment-1275344179) type in higher level languages.

### References

- https://github.com/hashicorp/terraform-cdk/issues/430
- https://github.com/hashicorp/terraform-cdk/issues/2001

### PR DESCRIPTION
**Pull request by ansgarm**
> Note! This PR is based on #3272 

## Problem
Terraform does not support iterating over lists without turning them into a set first. Sets that contain computed values need to be applied using `-target` in order for Terraform to know about these values before a plan is made. Using `-target` and applying changes in two passes is not desirable. However, there’s a workaround by turning them into maps which is described below.

### Workaround by using Maps
A workaround for this case is to turn the set of objects into a map with keys that are based on arguments that are known during the plan stage. This is done for example in the [HCL docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/acm_certificate) snippet for the `aws_acm_certificate_validation` resource:

image

In the snippet above, the set of objects stored in `domain_validation_options` is converted into a map with the `domain_name` of each object as the key.

If one instead tries to directly use `domain_validation_options`, the following error is printed (hinting at using `-target`, as described in the problem section above):

image

## Solution
Allow the user to pass the attribute name to be used as the key in the map and convert the set/list into a map using a for expression.

Closes #2178
Related to #2001, #1393, #430 (some already closed with workarounds)

### CODE REVIEW

File: packages/cdktf/lib/terraform-iterator.ts [hunk 1]
```diff
@@ -66,6 +70,22 @@ export abstract class TerraformIterator implements ITerraformIterator {
...
   }
 
+  /**
+   * Creates a new iterator from a complex list. One example for this would be a list of maps.
+   * The list will be converted into a map with the mapKeyAttributeName as the key.
+   * @param list the list to iterate over
+   * @param mapKeyAttributeName the name of the attribute that should be used as the key in the map
+   *
+   * @example
+   * TODO: put AWS certificate validation example
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/terraform-iterator.ts:80]: Leftover TODO :)

File: packages/cdktf/lib/terraform-iterator.ts [hunk 2]
```diff
@@ -66,6 +70,22 @@ export abstract class TerraformIterator implements ITerraformIterator {
...
+   */
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/terraform-iterator.ts:81]: Maybe we can also link to the not-yet-existing documentation page, I think the topic is complex enough that it warrants a new section in some page

File: test/typescript/iterators/test.ts [hunk 3]
```diff
@@ -60,6 +60,18 @@ describe("iterators integration test", () => {
...
+
+    expect(stack2.byId("record")).toHaveProperty(
+      "for_each",
+      "${{ for key, val in tolist(aws_acm_certificate.cert.domain_validation_options): val.domain_name => val }}"
+    );
+    expect(stack2.byId("record")).toHaveProperty("name", "${each.value.name}");
+  });
+
+  test("non-synth stack is valid Terraform", async () => {
+    await driver.exec("terraform", ["init"]);
```
Comment (DanielMSchmidt) [inline test/typescript/iterators/test.ts:72]: Nit: maybe the init & validate combination could be a new driver method

**Review by DanielMSchmidt — COMMENTED** [review]
I like the approach overall and to my knowledge it looks good 👍 There are a few things missing for the overall picture from my point of view, let me know what you think:

- [ ] Documentation section about this (detailing the use-case & usage)
- [ ] Hcl2cdk needs to be made aware of this, ideally we would detect dynamic blocks / for_each being used from a complex list and then add the correct fromComplexList invocation. This will allow us to remove yet another escape hatch 🎉

**Comment by ansgarm** [comment]
> I like the approach overall and to my knowledge it looks good 👍 There are a few things missing for the overall picture from my point of view, let me know what you think:
> 
>     * [ ]  Documentation section about this (detailing the use-case & usage)
> 
>     * [ ]  Hcl2cdk needs to be made aware of this, ideally we would detect dynamic blocks / for_each being used from a complex list and then add the correct fromComplexList invocation. This will allow us to remove yet another escape hatch 🎉

Added docs for this and linked them in the docstring as well.

In regards to `hcl2cdk` support: While it might be possible to detect for expression being used to convert set data structures containing computed values, I would put this out of scope for this PR as that seems a bit too complex to add onto this one.

File: examples/python/documentation/Pipfile [hunk 4]
```diff
@@ -7,4 +7,4 @@ verify_ssl = true
 python_version = "3.7"
 
 [packages]
-cdktf = "*"
+cdktf = {path = "./../../../dist/python/cdktf-0.19.1.dev1700818458-py3-none-any.whl"}
```
Comment (DanielMSchmidt) [inline examples/python/documentation/Pipfile:10]: This change is probably not wanted (and I think if you use the reinstall command in the package.json not needed)

**Review by DanielMSchmidt — APPROVED** [review]

File: examples/python/documentation/Pipfile (hunk 4 above)
Comment (ansgarm) [reply examples/python/documentation/Pipfile:10]: Whoops, yeah, that must've happened automatically somehow

**Comment by ansgarm** [comment]
Merging even though C# example fails on the `pluckProperty` call, which is a problem that's not directly related to this branch and something that @DanielMSchmidt is looking into.

**Pull request merged**